libbeat/cmd/instance/beat.go

```diff
@@ -696,8 +707,69 @@ func (b *Beat) registerTemplateLoading() error {
 			return err
 		}
 
-		if esCfg.Index != "" && (cfg.Name == "" || cfg.Pattern == "") && (b.Config.Template == nil || b.Config.Template.Enabled()) {
-			return fmt.Errorf("setup.template.name and setup.template.pattern have to be set if index name is modified.")
+		if esCfg.Index != "" && (templateCfg.Name == "" || templateCfg.Pattern == "") && (b.Config.Template == nil || b.Config.Template.Enabled()) {
+			return fmt.Errorf("setup.template.name and setup.template.pattern have to be set if index name is modified")
+		}
+
+		if b.Config.ILM.Enabled() {
+
+			cfgwarn.Beta("Index lifecycle management is enabled which is in beta.")
+			logp.Info("Overwriting setup.template for ILM")
+
+			ilmCfg, err := getILMConfig(b)
+			if err != nil {
+				return err
+			}
+
+			// Template name and pattern can't be configure when using ILM
+			logp.Info("Overwriting template name and pattern as ILM is enabled.")
+			b.Config.Template.SetString("name", -1, ilmCfg.RolloverAlias)
+			b.Config.Template.SetString("pattern", -1, fmt.Sprintf("%s-*", ilmCfg.RolloverAlias))
+
+			// rollover_alias and lifecycle.name can't be configured and will be overwritten
+			logp.Info("Overwrite template settings for rollover_alias and lifecycle.name.")
+			err = b.Config.Template.SetString("settings.index.lifecycle.rollover_alias", -1, ilmCfg.RolloverAlias)
+			if err != nil {
+				logp.Err("error setting rollover_alias setting: %s", err)
+			}
+			err = b.Config.Template.SetString("settings.index.lifecycle.name", -1, ILMPolicyName)
+			if err != nil {
+				logp.Err("error setting index lifecycle name setting err: %s", err)
+			}
+
+			// Set the ingestion index to the rollover alias
+			logp.Info("Overwrite index setting with rollover_alias: %s", ilmCfg.RolloverAlias)
+			esCfg.Index = ilmCfg.RolloverAlias
```

Comment:
What if `index` and `indices` were configured? Does this mean that you cannot use ILM and rollover when writing to multiple indices from a beat?

Reply:
It's an issue that bugs me and I don't have good answer for it yet. The problem:

Assuming we allow index alias to be based on fields of events like we do at the moment for indices (for example APM ...), how do we know when to create an index alias? It would mean we create multiple index alias but not at startup as we don't know yet on startup. At the same time we need to know if an index alias already exist to make sure we create it before the first event is created. We could keep some state on the Beats side to figure out which aliases are already created and exist. An alternative could be to disable automatic index creation: https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#index-creation But I think this would require a config change. There are some discussions to allow this on a cluster level. Assuming this would be enabled, we would get an error back and could create the index write alias.

Other ideas / thoughts?

To keep complexity low I would also be ok to ship a first version of ILM without this capability as it could be manually configured the right way assuming something creates the correct write aliases in advance. In the APM case it's 3 predefined ones, so it could even be hardcoded.

Reply:
We will need to be able to use multiple indices for APM Server. I am fine with dealing with that in a separate PR though.

Reply:
This is a long standing issue. Settings in outputs allow for multiple indices, but setup is focused on one index only. I don't think we can't fix it here.

Ultimately I would like to combine index selection with index/template/ilm setup. Right now a many settings are all over the place + must be configured appropriately.

Reply:
I put up a WIP PR to allow having multiple templates per beat https://github.com/elastic/beats/pull/9247. 
We will need that for ILM in APM. I had discussions going on with @ruflin about this for some time. Please also give feedback on this WIP PR.